### PR TITLE
Limit fan-In during compaction.

### LINKED ISSUE #4857: Make the compaction process with the new storage layer hierarchical
We need to make the compaction process hierarchical to prevent excessive object storage request fan-out and memory usage. The current compaction process is susceptible to object storage throttling and/or getting OOM killed when a large number of filesets are getting compacted (this can occur because we will buffer a chunk from each fileset being compacted). To prevent this, we need some configurable bound on the fan-out during the compaction process that will be enforced by a hierarchical compaction scheme. The compaction code starts here https://github.com/pachyderm/pachyderm/blob/03f22b5ebec7f56813fced4f7215de46c50ccf20/src/server/pfs/server/driver_new_storage.go#L285. This should be pretty straightforward because we just need to perform the work in batches as we work up the hierarchy until we are left with the final compacted fileset. The fan-out should be exposed as a storage configuration here https://github.com/pachyderm/pachyderm/blob/03f22b5ebec7f56813fced4f7215de46c50ccf20/src/server/pkg/serviceenv/config.go#L69. With this change, we should also think a bit more about how we should be setting up the compaction shards, and what configuration should be available for this. Figuring out a good default for these things will be important for good performance out of the box.

### PR DESCRIPTION
**Pull request by brendoncarroll**
Limits the number of objects that are open at a time during compact.  Adds a configurable `StorageCompactFanIn` parameter.

The workers will never have to open more than the fan-in value.  The master may need to compact more than fan-in filesets at once, but they will be well clustered because of a previous sharded compaction, and they can be concatenated while maintaining a correct ordering of keys in the final output. During concatenation only one fileset needs to be read from at a time.

Closes #4857

### CODE REVIEW

File: src/server/pfs/server/driver_new_storage.go [hunk 1]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
 	return nil
 }
 
-func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes []string) (retErr error) {
+// compact is the entrypoint to compaction.
+func (d *driver) compact(ctx context.Context, master *work.Master, outputPath string, inputPrefixes []string) error {
+	// resolve prefixes into paths
+	inputPaths := []string{}
+	for _, inputPrefix := range inputPrefixes {
+		err := d.storage.WalkFileSets(ctx, inputPrefix, func(inputPath string) error {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:307]: This should be pulling from the master context (`m.Ctx()`), not the background. The `compact` function should not take a separate context as a parameter.

File: src/server/pfs/server/driver_new_storage.go [hunk 2]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
 	return nil
 }
 
-func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes []string) (retErr error) {
+// compact is the entrypoint to compaction.
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:302]: Kind of unnecessary comment.

File: src/server/pfs/server/driver_new_storage.go (hunk 1 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver_new_storage.go:307]: Also, `WalkFileSets` -> `WalkFileSet`

Comment (brycemcanally) [reply src/server/pfs/server/driver_new_storage.go:307]: Also, the if statement can just operate on the results of the function call, you don't need it to be below.

File: src/server/pfs/server/driver_new_storage.go [hunk 3]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
...
+	for _, inputPrefix := range inputPrefixes {
+		err := d.storage.WalkFileSets(ctx, inputPrefix, func(inputPath string) error {
+			inputPaths = append(inputPaths, inputPath)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	if d.env.StorageCompactionMaxFanIn < 2 {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:315]: We probably should not get to this point if this is the case, we probably want this grouped with other validation (I think we can punt on this tbh).

File: src/server/pfs/server/driver_new_storage.go [hunk 4]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
...
+	}
+	return d.compactIter(ctx, compactParams{
+		master:     master,
+		inputPaths: inputPaths,
+		outputPath: outputPath,
+		maxFanIn:   d.env.StorageCompactionMaxFanIn,
+	})
+}
+
+type compactParams struct {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:326]: I think `compactSpec` might be a better name.

File: src/server/pfs/server/driver_new_storage.go [hunk 5]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
...
+// compactIter is one level of compaction.  It will only perform compaction
+// if len(inputPaths) < params.maxFanIn otherwise it will split inputPaths recursively.
+func (d *driver) compactIter(ctx context.Context, params compactParams) (retErr error) {
+	scratch := path.Join(tmpPrefix, uuid.NewWithoutDashes())
+	defer func() {
+		if err := d.storage.Delete(ctx, scratch); retErr == nil {
+			retErr = err
+		}
+	}()
+	// exceeds maxFanIn need to branch out
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:342]: Might want to reword a bit.

File: src/server/pfs/server/driver_new_storage.go (hunk 4 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_new_storage.go:326]: Isn't there a method `CompactSpec` on the storage type that would get confused with this?

File: src/server/pfs/server/driver_new_storage.go [hunk 6]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
...
+	inputPaths := params.inputPaths
+	if len(inputPaths) > params.maxFanIn {
+		childOutputPaths := []string{}
+		childSize := len(params.inputPaths) / params.maxFanIn
+		if len(params.inputPaths)%params.maxFanIn != 0 {
+			childSize++
+		}
+		group, ctx := errgroup.WithContext(ctx)
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:350]: Just for consistency, `group` -> `eg`.

File: src/server/pfs/server/driver_new_storage.go (hunk 1 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_new_storage.go:307]: I always find it hard to read the compound if statements when they are broken up by a function literal.  What do you think?

File: src/server/pfs/server/driver_new_storage.go (hunk 3 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_new_storage.go:315]: I guess I can remove it. It will probably OOM pretty quickly anways.

File: src/server/pfs/server/driver_new_storage.go [hunk 7]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
...
+func (d *driver) compactIter(ctx context.Context, params compactParams) (retErr error) {
+	scratch := path.Join(tmpPrefix, uuid.NewWithoutDashes())
+	defer func() {
+		if err := d.storage.Delete(ctx, scratch); retErr == nil {
+			retErr = err
+		}
+	}()
+	// exceeds maxFanIn need to branch out
+	inputPaths := params.inputPaths
+	if len(inputPaths) > params.maxFanIn {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:344]: Something that I think could be beneficial from a readability standpoint would be to structure this how most recursive functions are where the base case is checked first, then the recursive step is handled after the check. So, you would basically invert what you have now, so the check would be less than or equal to fan in, then run sharded compact. If not, then kick off the recursion and run sharded compact on the results. Just an idea, what are your thoughts?

File: src/server/pfs/server/driver_new_storage.go [hunk 8]
```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
...
+		}
+		inputPaths = childOutputPaths
+	}
+	// within maxFanIn, generate compaction tasks for workers
+	return d.shardedCompact(ctx, params.master, params.outputPath, inputPaths)
+}
+
+// shardedCompact generates shards for the fileset(s) in inputPaths,
+// gives those shards to workers, and waits for them to complete
+// fan-in: spatially, and temporally to each shard will be len(inputPaths)
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:379]: Not sure I understand what these two lines are saying? Maybe reword? Fan in is bound by len(inputPaths), concatenating shards have fan in of one because they are concatenated sequentially.

File: src/server/pfs/server/driver_new_storage.go [hunk 9]
```diff
@@ -333,7 +415,22 @@ func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes [
 	}); err != nil {
 		return err
 	}
-	return d.storage.Compact(master.Ctx(), outputPath, []string{scratch})
+	return d.concatFileSets(ctx, outputPath, shardOutputs)
+}
+
+// concatFileSets concatenates the filesets in inputPaths and writes the result to outputPath
+// TODO: move this to the fileset package, and error if the entries are not sorted.
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:422]: :+1:

File: src/server/pfs/server/driver_new_storage.go (hunk 3 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver_new_storage.go:315]: Yeah, I think it is worth making a todo (and you could even have the todo somewhere around here), but I think this type of validation should be handled at setup, not while we are in the middle of a compaction.

File: src/server/pfs/server/driver_new_storage.go (hunk 4 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver_new_storage.go:326]: Yeah, that is actually why I prefer it. There is a notion of a fileset compact spec and a pfs compact spec. The fileset compact spec shards across a set of filesets, and the pfs compact spec manages fan out.

File: src/server/pfs/server/driver_new_storage.go (hunk 7 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_new_storage.go:344]: Yeah that could work.  I think it prevents the reassign to inputPaths, which is cleaner.  It also lets us skip the scratch space for the base case.  I'll give it a try.

File: src/server/pfs/server/driver_new_storage.go (hunk 1 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver_new_storage.go]: I kind of prefer this pattern for the simple case where I am just returning the error if it is not nil, but for more complex logic it could make sense to split it up. The general pattern in the codebase for the simple case is to just have the function call in the if statement.

File: src/server/pfs/server/driver_new_storage.go [hunk 10]
```diff
@@ -298,24 +299,99 @@ func (fr *FileReader) drain() error {
...
+	master     *work.Master
+	outputPath string
+	inputPaths []string
+	maxFanIn   int
+}
+
+// compactIter is one level of compaction.  It will only perform compaction
+// if len(inputPaths) < params.maxFanIn otherwise it will split inputPaths recursively.
+func (d *driver) compactIter(ctx context.Context, params compactSpec) (retErr error) {
+	// within maxFanIn
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:333]: Kind of unnecessary.

File: src/server/pfs/server/driver_new_storage.go [hunk 11]
```diff
@@ -298,24 +299,99 @@ func (fr *FileReader) drain() error {
...
+
+type compactSpec struct {
+	master     *work.Master
+	outputPath string
+	inputPaths []string
+	maxFanIn   int
+}
+
+// compactIter is one level of compaction.  It will only perform compaction
+// if len(inputPaths) < params.maxFanIn otherwise it will split inputPaths recursively.
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:331]: <=

File: src/server/pfs/server/testing/server_new_storage_test.go [hunk 12]
```diff
@@ -35,6 +37,7 @@ type loadConfig struct {
 func newLoadConfig(opts ...loadConfigOption) *loadConfig {
 	config := &loadConfig{}
 	config.pachdConfig = newPachdConfig()
+	config.pachdConfig.StorageCompactionMaxFanIn = 2
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_new_storage_test.go:40]: It seems like if this should be the default it would be in the `newPachdConfig` function, but this seems like a pretty low default.

File: src/server/pfs/server/driver_new_storage.go [hunk 13]
```diff
@@ -333,7 +409,30 @@ func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes [
 	}); err != nil {
 		return err
 	}
-	return d.storage.Compact(master.Ctx(), outputPath, []string{scratch})
+
+	var shardOutputs []string
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:413]: I think we should follow the same pattern you used for recording the `childOutputPaths` here. This does not actually preserve the shard order because walk will return the paths lexicographically.

File: src/server/pfs/server/testing/server_new_storage_test.go (hunk 12 above)
Comment (brendoncarroll) [reply src/server/pfs/server/testing/server_new_storage_test.go:40]: It should be 2 here because its the best number for the low memory testing, and least likely to explode CI.  The best default depends on the amount of memory we want to use. I think larger will almost always be better as far as time is concerned.

File: src/server/pfs/server/driver_new_storage.go (hunk 13 above)
Comment (brendoncarroll) [reply src/server/pfs/server/driver_new_storage.go:413]: I think this can produce an error if no shard is written out.

File: src/server/pfs/server/testing/server_new_storage_test.go (hunk 12 above)
Comment (brycemcanally) [reply src/server/pfs/server/testing/server_new_storage_test.go:40]: I think it makes sense for the default to be a bit higher, and for the low memory testing you should configure it as two with a functional option for that field. Also, the default should be set with the other defaults in `newPachdConfig`.

File: src/server/pfs/server/driver_new_storage.go (hunk 13 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver_new_storage.go:413]: If a shard is created, then a fileset, empty or not, should get written to the semantic path for a shard (if it is not, then there is a bug somewhere).

File: src/server/pfs/server/driver_new_storage.go [hunk 14]
```diff
@@ -429,6 +418,7 @@ func (d *driver) concatFileSets(ctx context.Context, outputPath string, inputPat
 		if err := fsr.Iterate(func(fr *fileset.FileReader) error {
 			return fsw.CopyFile(fr)
 		}); err != nil {
+			panic(err)
```
Comment (brycemcanally) [inline src/server/pfs/server/driver_new_storage.go:421]: Remove panic debug.
